### PR TITLE
Document the new gulp.task API.

### PR DESCRIPTION
**Pull request by dinoboff**
Also add the gulp.series and gulp.parallel APIs.

#803, #841 and #925

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -104,132 +105,183 @@ gulp.task('somename', function() {
...
+});
 
-2. In task "two" you add a hint telling the engine that it depends on completion of the first task.
+```
 
-So this example would look like this:
+**Note:** By default, tasks run with maximum concurrency -- e.g. it launches 
+all the tasks at once and waits for nothing.
 
+To run serial tasks, use `gulp.serial` with asynchronous tasks:
```
Comment (heikki) [inline docs/API.md]: gulp.series

File: docs/API.md [hunk 2]
```diff
@@ -104,132 +105,183 @@ gulp.task('somename', function() {
...
+```js
+gulp.task('clean', function() {
+  return spawn('rm', ['-rf', path.join(__dirname, 'build')]);
+});
 
-2. In task "two" you add a hint telling the engine that it depends on completion of the first task.
+```
 
-So this example would look like this:
+**Note:** By default, tasks run with maximum concurrency -- e.g. it launches
```
Comment (heikki) [inline docs/API.md]: Is this note still valid?

File: docs/API.md [hunk 3]
```diff
@@ -104,132 +105,183 @@ gulp.task('somename', function() {
...
+}]));
+```
 
-#### opts
-Type: `Object`
+#### tasks
+Type: `Array`, `String` or `Function`
 
-Options, that are passed to [`gaze`](https://github.com/shama/gaze).
+A task name, a functions or an array of either.
```
Comment (heikki) [inline docs/API.md]: function

File: docs/API.md [hunk 4]
```diff
@@ -104,132 +105,183 @@ gulp.task('somename', function() {
...
 
-#### tasks
-Type: `Array`
 
-Names of task(s) to run when a file changes, added with `gulp.task()`
+### gulp.series(tasks, ...)
+
+Takes a variable amount of task names or function and returns a function
```
Comment (heikki) [inline docs/API.md]: functions

File: docs/API.md [hunk 5]
```diff
@@ -104,132 +105,183 @@ gulp.task('somename', function() {
...
+  // do more stuff
+  done();
+}]));
 ```
 
-### gulp.watch(glob[, opts, cb])
+#### tasks
+Type: `Array`, `String` or `Function`
+
+A task name, a functions or an array of either.
```
Comment (heikki) [inline docs/API.md]: function

File: docs/API.md (hunk 2 above)
Comment (dinoboff) [inline docs/API.md]: You're right. It's not relevant in the `gulp.task` documentation

File: docs/API.md [hunk 6]
```diff
@@ -92,9 +92,10 @@ Default: `0777`
 
 Octal permission string specifying mode for any folders that need to be created for output folder.
 
-### gulp.task(name[, deps], fn)
 
-Define a task using [Orchestrator].
+### gulp.task([name,] fn)
+
+Define a task using [Undertaker].
```
Comment (phated) [inline docs/API.md]: should link to undertaker

File: docs/API.md [hunk 7]
```diff
@@ -104,132 +105,161 @@ gulp.task('somename', function() {
...
 
-- give it a hint to tell it when the task is done,
-- and give it a hint that a task depends on completion of another.
+gulp.task('clean', function() {
+  return promisedDel(['.build/']);
+});
+```
 
-For these examples, let's presume you have two tasks, "one" and "two" that you specifically want to run in this order:
+##### Return a child process
```
Comment (phated) [inline docs/API.md]: we also except RxJS observables now

File: docs/API.md (hunk 6 above)
Comment (dinoboff) [inline docs/API.md]: It is; it's a link and its URL is defined at the end of the file.

File: docs/API.md [hunk 8]
```diff
@@ -104,132 +105,161 @@ gulp.task('somename', function() {
...
+var del = require('del');
 
-gulp.task('somename', function() {
-  var deferred = Q.defer();
+gulp.task('clean', function(done) {
+  del(['.build/'], function(err) {
+    done(err)
+  });
+  // or
+  // del(['.build/'], done);
```
Comment (phated) [inline docs/API.md]: can we just use this example instead of the extra functions.  keep things more performant in the docs

File: docs/API.md [hunk 9]
```diff
@@ -104,132 +105,161 @@ gulp.task('somename', function() {
...
-2. In task "two" you add a hint telling the engine that it depends on completion of the first task.
+```
 
-So this example would look like this:
 
-```js
-var gulp = require('gulp');
+### gulp.parallel(tasks, ...)
+
+Takes a variable amount of task names or function and returns
```
Comment (phated) [inline docs/API.md]: need to mention that if you use task names, tasks must be registered prior to calling gulp.parallel (same goes for series)

File: docs/API.md [hunk 10]
```diff
@@ -104,132 +105,161 @@ gulp.task('somename', function() {
...
+A task name, a function or an array of either.
 
-Watch files and do something when a file changes. This always returns an EventEmitter that emits `change` events.
 
-### gulp.watch(glob[, opts], tasks)
+### gulp.series(tasks, ...)
 
-#### glob
-Type: `String` or `Array`
+Takes a variable amount of task names or function and returns a function
```
Comment (phated) [inline docs/API.md]: see note on parallel

File: docs/API.md [hunk 11]
```diff
@@ -241,33 +271,26 @@ Type: `Object`
 
 Options, that are passed to [`gaze`](https://github.com/shama/gaze).
 
-#### cb(event)
-Type: `Function`
+#### tasks
+Type: `Array`, `Function` or `String`
 
-Callback to be called on each change.
+A task name, a function or an array of either to run when a file changes.
```
Comment (phated) [inline docs/API.md]: make not about the reason to use `watcher.on('change')` if the gaze `event` is needed

File: docs/getting-started.md [hunk 12]
```diff
@@ -3,7 +3,7 @@
 #### 1. Install gulp globally:
 
 ```sh
-$ npm install --global gulp
+$ npm install --global gulp-cli
```
Comment (phated) [inline docs/getting-started.md]: hold off on this until we get the blockers worked through

File: docs/API.md [hunk 13]
```diff
@@ -104,132 +105,161 @@ gulp.task('somename', function() {
...
+bar.name = 'bar'
 ```
 
-**Note:** Are your tasks running before the dependencies are complete?  Make sure your dependency tasks are correctly using the async run hints: take in a callback or return a promise or event stream.
+You should either provide the task name or avoid anonymous functions.
 
 #### fn
 
-The function that performs the task's operations. Generally this takes the form of `gulp.src().pipe(someplugin())`.
+The function that performs the task's operations.
```
Comment (phated) [inline docs/API.md]: note that these __cannot__ be synchronous anymore

File: docs/API.md (hunk 7 above)
Comment (dinoboff) [inline docs/API.md]: Do you have an example?

File: docs/API.md (hunk 6 above)
Comment (phated) [inline docs/API.md]: oh, okay

File: docs/API.md (hunk 7 above)
Comment (phated) [inline docs/API.md]: https://github.com/phated/async-done/blob/master/test/observables.js

File: docs/recipes/running-tasks-in-series.md [hunk 14]
```diff
@@ -1,66 +1,65 @@
...
-
-For these examples, let's presume you have two tasks, "one" and "two" that you specifically want to run in this order:
-
-1. In task "one" you add a hint to tell it when the task is done. Either take in a callback and call it when you're done or return a promise or stream that the engine should wait to resolve or end respectively.
-
-2. In task "two" you add a hint telling the engine that it depends on completion of the first task.
-
-So this example would look like:
+- use `gulp.series`;
+- and your tasks should give hint of async operation by returning a stream,
```
Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: and RxJS observable

File: docs/recipes/running-tasks-in-series.md [hunk 15]
```diff
@@ -1,66 +1,65 @@
...
+  // if err is not null and not undefined, the orchestration will stop,
+  // and 'two' will not run
+  done(err); 
 });
 
 // identifies a dependent task must be complete before this one begins
-gulp.task('two', ['one'], function() {
-    // task 'one' is done now
+gulp.task('two', gulp.series('one', function() {
+  // task 'one' is done now
```
Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: all tasks are async

**Comment by phated** [comment]
@dinoboff you are amazing. thanks for all the hard work

File: docs/API.md [hunk 16]
```diff
@@ -104,132 +105,157 @@ gulp.task('somename', function() {
 
 #### name
 
-The name of the task. Tasks that you want to run from the command line should not have spaces in them.
+Optional, The name of the task. Tasks that you want to run from the command line 
+should not have spaces in them.
 
-#### deps
-Type: `Array`
+If the name is not provided, the task will be name after the function `name` attribute.
```
Comment (contra) [inline docs/API.md]: should note that named functions will have a .name attribute

File: docs/API.md [hunk 17]
```diff
@@ -104,132 +105,157 @@ gulp.task('somename', function() {
...
+  done();
+});
 
-Options, that are passed to [`gaze`](https://github.com/shama/gaze).
+gulp.task('two', function(done) {
+  // do stuff
+  done();
+});
+
+gulp.task('default', gulp.series(['one', 'two', function(done) {
```
Comment (contra) [inline docs/API.md]: are we going to show array or function arity usage in the docs? We have to pick one we can't use both. I vote for the variable function arity usage

File: docs/API.md [hunk 18]
```diff
@@ -104,132 +105,157 @@ gulp.task('somename', function() {
...
+  // do more stuff
+  done();
+}]));
+```
 
 #### tasks
```
Comment (contra) [inline docs/API.md]: should we really support this?

```js
gulp.watch('glob', ['one', 'two', 'three'])
```

Wouldn't it be easier to just tell people to provide a task fn as the second argument

```js
gulp.watch('glob', gulp.parallel('one', 'two', 'three'))
```

File: docs/recipes/running-tasks-in-series.md (hunk 15 above)
Comment (dinoboff) [inline docs/recipes/running-tasks-in-series.md]: So all tasks are async, they must either return async-y value use the callback argument?

File: docs/API.md (hunk 18 above)
Comment (phated) [inline docs/API.md]: @contra I believe we discussed this in the past and decided it wasn't worth breaking the syntax on watch and the array would just execute in parallel.  I remember rebasing to add the functionality back in after I removed.  I'd be fine with removing the array syntax if we want to change the decision (it simplifies .watch).

Comment (heikki) [inline docs/API.md]: It would be good if `gulp.watch` and `gulp.task` had similar syntaxes.

File: docs/API.md (hunk 16 above)
Comment (dinoboff) [inline docs/API.md]: I am changing the note to:

    `Function.name` is not writable; it cannot be set or edited.
    It will be empty for anonymous function:

File: docs/API.md [hunk 19]
```diff
@@ -104,132 +105,170 @@ gulp.task('somename', function() {
...
+var Observable = require('rx').Observable;
 
-// takes in a callback so the engine knows when it'll be done
-gulp.task('one', function(cb) {
-    // do stuff -- async or otherwise
-    cb(err); // if err is not null and not undefined, the run will stop, and note that it failed
+gulp.task('sometask', function() {
+  return Observable.return(42);
 });
+```
```
Comment (dinoboff) [inline docs/API.md]: @phated is it ok?

File: docs/API.md (hunk 11 above)
Comment (dinoboff) [inline docs/API.md]: Can I leave that for #925 ?

File: docs/API.md (hunk 17 above)
Comment (dinoboff) [inline docs/API.md]: I was mixing them, It should now only show arity usage.

File: docs/recipes/running-tasks-in-series.md (hunk 15 above)
Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: correct.  If you don't do either of those, the task hangs forever

File: docs/API.md (hunk 13 above)
Comment (dinoboff) [inline docs/API.md]: Changed. there shouldn't be any sync task left.

File: docs/API.md [hunk 20]
```diff
@@ -104,132 +105,175 @@ gulp.task('somename', function() {
...
 
+The callback accept an optional `Error` object. If it receives an error, 
+the task will fail.
+
 ##### Return a stream
 
 ```js
 gulp.task('somename', function() {
-  var stream = gulp.src('client/**/*.js')
+  return gulp.src('client/**/*.js')
```
Comment (phated) [inline docs/API.md]: can we make sure all examples have this update?  I like it much more than `var stream = ...; return stream`

**Comment by dinoboff** [comment]
Thanks.

It covers `gulp.task` and `gulp.watch`. 

I don't know if there is any changes to `vinyl-fs` that need to be documented.

File: docs/recipes/running-tasks-in-series.md [hunk 21]
```diff
@@ -1,66 +1,61 @@
...
 ```js
 var gulp = require('gulp');
 
 // takes in a callback so the engine knows when it'll be done
-gulp.task('one', function(cb) {
-    // do stuff -- async or otherwise
-    cb(err); // if err is not null and not undefined, the orchestration will stop, and 'two' will not run
+gulp.task('one', function(done) {
+  // do stuff -- async or otherwise
+  // if err is not null and not undefined, the orchestration will stop,
```
Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: remove the work "orchestration" throughout the docs

File: docs/recipes/running-tasks-in-series.md [hunk 22]
```diff
@@ -1,66 +1,61 @@
...
+It also uses `once` to run dependencies once:
 ```js
 var gulp = require('gulp');
-var del = require('del'); // rm -rf
+var promisedDel = require('promised-del'); // rm -rf
+var once = require('once');
+
+gulp.task('clean', once(function() {
+  // The clean task will be run twice, but this function will be run once.
+  // So we cannnot rely on a callback.
```
Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: why can't we rely on the callback?  Is this a limitation of `once`?

File: docs/recipes/running-tasks-in-series.md [hunk 23]
```diff
@@ -1,66 +1,61 @@
...
+  // Instead we return a promise; the same promise will be returned twice.
+  return promisedDel(['output']);
+}));
+
+gulp.task('templates', gulp.series('clean', function() {
+  var stream = gulp.src(['src/templates/*.hbs'])
```
Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: switch to the `return gulp.src` pattern

File: docs/recipes/running-tasks-in-series.md [hunk 24]
```diff
@@ -1,66 +1,61 @@
...
-    del(['output'], cb);
 });
 
-gulp.task('templates', ['clean'], function() {
-    var stream = gulp.src(['src/templates/*.hbs'])
-        // do some concatenation, minification, etc.
-        .pipe(gulp.dest('output/templates/'));
-    return stream; // return the stream as the completion hint
+gulp.task('styles', gulp.series('clean', function() {
+  var stream = gulp.src(['src/styles/app.less'])
```
Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: see note above

File: docs/recipes/running-tasks-in-series.md (hunk 21 above)
Comment (dinoboff) [inline docs/recipes/running-tasks-in-series.md]: It was the only occurrence.

File: docs/recipes/running-tasks-in-series.md [hunk 25]
```diff
@@ -1,66 +1,59 @@
...
-var del = require('del'); // rm -rf
+var promisedDel = require('promised-del'); // rm -rf
+var once = require('once');
 
-gulp.task('clean', function(cb) {
-    del(['output'], cb);
-});
+gulp.task('clean', once(function() {
+  // The clean task will be run twice, but this function will be run once.
+  // So we cannnot rely on a callback.
```
Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: why can't we rely on the callback? Is this a limitation of once? (added to the non-outdated diff)

File: docs/recipes/running-tasks-in-series.md (hunk 22 above)
Comment (dinoboff) [inline docs/recipes/running-tasks-in-series.md]: `once` will not run the function a second time, the callback will never be called and the task keep going on the second time it's called.

File: docs/recipes/running-tasks-in-series.md (hunk 25 above)
Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: just dug into this and it seems to exist in both `once` and [`onetime`](https://github.com/sindresorhus/onetime) - I think we need an async version of `once` so it doesn't just rely on returning a promise

Comment (dinoboff) [inline docs/recipes/running-tasks-in-series.md]: `once` will not run the function a second time, just send the return value of the first call for any other call.

Using the callback pattern, for the second call the callback will never be called and the task will keep going on.

Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: Just published a module that should solve that problem - https://www.npmjs.com/package/async-once

Please call that out as the solution to use with callbacks + an example

Comment (heikki) [inline docs/recipes/running-tasks-in-series.md]: Is separate "Running tasks in series" recipe needed now that we have series / parallel? I'd also like to see all the references to "dependency" removed because there is no such thing anymore. Recipe for "Running function only once" might be useful for some, though I'd see that just as a sign of thinking the problem wrong way around.

Comment (phated) [inline docs/recipes/running-tasks-in-series.md]: I think we could do away with the series recipe contents and just add a link that points to the docs for `series`.

I agree about doing away with the "dependency" terminology because everything is about function composition now.

I think we should have a "running once" recipe because people have come to expect that behavior from current gulp.  The recipe just makes the upgrade path nicer.

Comment (dinoboff) [inline docs/recipes/running-tasks-in-series.md]: Updated:
- the example uses async-once.
- removed the dependency terminology, except that's I am calling the example using once a "dependency pattern".

**Comment by callumacrae** [comment]
I can proofread this when it's done :smile:

**Comment by dinoboff** [comment]
@callumacrae thanks; please do.

File: docs/API.md [hunk 26]
```diff
@@ -2,46 +2,56 @@
 
 ### gulp.src(globs[, options])
 
-Emits files matching provided glob or an array of globs. 
-Returns a [stream](http://nodejs.org/api/stream.html) of [Vinyl files](https://github.com/wearefractal/vinyl-fs) 
-that can be [piped](http://nodejs.org/api/stream.html#stream_readable_pipe_destination_options) 
-to plugins.
+Emits files matching provided glob or an array of globs.
```
Comment (callumacrae) [inline docs/API.md]: I would get rid of the "an" here, it makes it more obvious that the files are matching the array of globs, not that an array of globs is emitted.

File: docs/API.md [hunk 27]
```diff
@@ -2,46 +2,56 @@
...
+```
+
 #### options
 Type: `Object`
 
 Options to pass to [node-glob] through [glob-stream].
 
-gulp adds some additional options in addition to the [options supported by node-glob][node-glob documentation] and [glob-stream]:
+gulp adds some additional options in addition to the
+[options supported by node-glob][node-glob documentation] and [glob-stream]:
```
Comment (callumacrae) [inline docs/API.md]: I'm assuming the line breaks you're introducing into this file are to keep line lengths low?

File: docs/API.md [hunk 28]
```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
...
+Optional, The name of the task. Tasks that you want to run from the command line
+should not have spaces in them.
 
-#### deps
-Type: `Array`
+If the name is not provided, the task will be name after the function `name` attribute.
 
-An array of tasks to be executed and completed before your task will run.
+`Function.name` is not writable; it cannot be set or edited.
+It will be empty for anonymous function:
```
Comment (callumacrae) [inline docs/API.md]: functions plural

File: docs/API.md [hunk 29]
```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
...
+You should either provide the task name or avoid anonymous functions.
 
 #### fn
 
-The function that performs the task's operations. Generally this takes the form of `gulp.src().pipe(someplugin())`.
+The function that performs the task's operations.
+Generally this takes the form of `return gulp.src().pipe(someplugin());`.
 
-#### Async task support
+Gulp assumes each task is asynchronous and call them with a callback parameter.
```
Comment (callumacrae) [inline docs/API.md]: call -> calls

File: docs/API.md [hunk 30]
```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
...
+A Task should either call the callback parameter to signal when operations
```
Comment (callumacrae) [inline docs/API.md]: Task -> task, lowercase t

File: docs/API.md [hunk 31]
```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
...
-  });
+```js
+var del = require('del');
+
+gulp.task('clean', function(done) {
+  del(['.build/'], done);
 });
 ```
 
+The callback accept an optional `Error` object. If it receives an error,
```
Comment (callumacrae) [inline docs/API.md]: accept -> accepts

File: docs/API.md [hunk 32]
```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
...
-  return deferred.promise;
+```js
+gulp.task('clean', function() {
+  return spawn('rm', ['-rf', path.join(__dirname, 'build')]);
 });
+
 ```
 
-**Note:** By default, tasks run with maximum concurrency -- e.g. it launches all the tasks at once and waits for nothing. If you want to create a series where tasks run in a particular order, you need to do two things:
+##### Return a RxJS observable
```
Comment (callumacrae) [inline docs/API.md]: Is it work putting a link to RxJS here somewhere? Most people haven't heard of it.

File: docs/API.md [hunk 33]
```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
...
+gulp.task('sometask', function() {
+  return Observable.return(42);
+});
+```
 
-For these examples, let's presume you have two tasks, "one" and "two" that you specifically want to run in this order:
 
-1. In task "one" you add a hint to tell it when the task is done.  Either take in a callback and call it when you're
-done or return a promise or stream that the engine should wait to resolve or end respectively.
+### gulp.parallel(tasks, ...)
```
Comment (callumacrae) [inline docs/API.md]: Use a [rest parameter](http://tc39wiki.calculist.org/es6/rest-parameters/)! :smile: 

`gulp.parallel(...tasks)`

File: docs/API.md [hunk 34]
```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
...
 
-2. In task "two" you add a hint telling the engine that it depends on completion of the first task.
+Takes a variable amount of task names or function and returns
+a function of the composed tasks or functions. When using task names,
+the task should already registered.
```
Comment (callumacrae) [inline docs/API.md]: This doesn't make sense. Should already _be_ registered?

File: docs/API.md [hunk 35]
```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
...
+A task name, a function or an array of either.
 
-Watch files and do something when a file changes. This always returns an EventEmitter that emits `change` events.
 
-### gulp.watch(glob[, opts], tasks)
+### gulp.series(tasks, ...)
 
-#### glob
-Type: `String` or `Array`
+Takes a variable amount of task names or function and returns
```
Comment (callumacrae) [inline docs/API.md]: "Takes a number of task names" reads better, imo

File: docs/API.md [hunk 36]
```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
...
+a function of the composed tasks or functions. When using task names,
+the tasks should already registered.
```
Comment (callumacrae) [inline docs/API.md]: same again, missing a word here

File: docs/API.md (hunk 31 above)
Comment (dinoboff) [inline docs/API.md]: @phated the callback signature is `callback(error, result)` I think. Does the result parameter matter in Gulp?

File: docs/API.md [hunk 37]
```diff
@@ -104,133 +177,185 @@ gulp.task('somename', function() {
...
 
 ```js
-var watcher = gulp.watch('js/**/*.js', ['uglify','reload']);
-watcher.on('change', function(event) {
-  console.log('File ' + event.path + ' was ' + event.type + ', running tasks...');
-});
+gulp.watch('js/**/*.js', ['uglify', 'reload', function(done) {
+  // do something
+  done();
+}]);
```
Comment (callumacrae) [inline docs/API.md]: This isn't my decision to make, but I feel it probably shouldn't be encouraged to pass functions directly into `watch`. Sure, mention it below, but don't do it in a code example.

@phated @contra: thoughts?

File: docs/API.md [hunk 38]
```diff
@@ -241,33 +366,47 @@ Type: `Object`
...
+gulp.watch('*.js', gulp.parallel(one, two));
+```
 
+`gulp.watch` returns an `EventEmitter` object which emits `change` events:
 ```js
-gulp.watch('js/**/*.js', function(event) {
+var watcher = gulp.watch('js/**/*.js', ['uglify', 'reload', function(done) {
+  // do something
+  done();
+}]);
```
Comment (callumacrae) [inline docs/API.md]: I feel adding the anonymous function here adds needless complexity

File: docs/recipes/running-tasks-in-series.md [hunk 39]
```diff
@@ -1,66 +1,95 @@
-# Running tasks in series, i.e. Task Dependency
+# Running tasks in series
 
-By default, tasks run with maximum concurrency -- e.g. it launches all the tasks at once and waits for nothing. If you want to create a series where tasks run in a particular order, you need to do two things:
+By default, gulp cli run task with maximum concurrency -- e.g. it launches all the tasks at once
```
Comment (callumacrae) [inline docs/recipes/running-tasks-in-series.md]: cli -> CLI?

"--" -> "—"?

Why do you have a line break at 96 characters? lol

File: docs/recipes/running-tasks-in-series.md [hunk 40]
```diff
@@ -1,66 +1,95 @@
...
+[`async-once`](https://www.npmjs.com/package/async-once) to run the `clean`
+task operations only once:
 ```js
 var gulp = require('gulp');
-
-// takes in a callback so the engine knows when it'll be done
-gulp.task('one', function(cb) {
-    // do stuff -- async or otherwise
-    cb(err); // if err is not null and not undefined, the orchestration will stop, and 'two' will not run
+var del = require('del'); // rm -rf
```
Comment (callumacrae) [inline docs/recipes/running-tasks-in-series.md]: Why the `// rm -rf`?

File: docs/recipes/running-tasks-in-series.md [hunk 41]
```diff
@@ -1,66 +1,95 @@
...
+var once = require('async-once');
+
+gulp.task('clean', once(function(done) {
+  // Run only once.
+  // For the next call to the clean task, once will call done with 
+  // the same arguments than the first call.
```
Comment (callumacrae) [inline docs/recipes/running-tasks-in-series.md]: than -> as

Comment (callumacrae) [inline docs/recipes/running-tasks-in-series.md]: also gulp uses lowercase letters at the beginning of comments, not sure if that's something you care about

File: docs/recipes/running-tasks-in-series.md [hunk 42]
```diff
@@ -1,66 +1,95 @@
...
+// even though it is called as a dependency twice.
+gulp.task('build', gulp.parallel('templates', 'styles'));
 
-Another example, which returns the stream instead of using a callback:
+// an alias.
+gulp.task('default', gulp.parallel('build'));
+```
 
-```js
+Note that it's an anti-pattern in Gulp 4 and the logs will show the clean task
```
Comment (callumacrae) [inline docs/recipes/running-tasks-in-series.md]: If it's an anti-pattern, why is it in the documentation?

**Comment by callumacrae** [comment]
That's everything! Most of my comments can be ignored, but I hope some of them are helpful :smile:

File: docs/API.md (hunk 37 above)
Comment (dinoboff) [inline docs/API.md]: I agree. I put it there to show it like other composed task/function. But the main example be more explicit anyway, and use `gulp.parallel`:
```js
gulp.watch('js/**/*.js', gulp.parallel('uglify', 'reload'));
```
An example with array it shown later.

File: docs/recipes/running-tasks-in-series.md (hunk 41 above)
Comment (dinoboff) [inline docs/recipes/running-tasks-in-series.md]: I didn't know.

File: docs/recipes/running-tasks-in-series.md (hunk 42 above)
Comment (dinoboff) [inline docs/recipes/running-tasks-in-series.md]: I might be better to put it in the upcoming migration guide.

File: docs/recipes/running-tasks-in-series.md (hunk 41 above)
Comment (dinoboff) [inline docs/recipes/running-tasks-in-series.md]: Updated; I checked, I think it's the only occurrence.

**Comment by dinoboff** [comment]
Thanks, they all made sense.

File: docs/API.md (hunk 20 above)
Comment (dinoboff) [inline docs/API.md]: checked.

File: docs/API.md (hunk 27 above)
Comment (dinoboff) [inline docs/API.md]: Yes, I didn't try to keep all lines under 80. I only cut the very long ones.

Comment (callumacrae) [inline docs/API.md]: It's a bit odd, imo, especially here:

https://github.com/gulpjs/gulp/pull/937/files#diff-96cc378dcdef180641176501f42183c7R286

Imo (and what I usually do) is either consistently break the lines all at 80 characters, or don't break them.

**Comment by dinoboff** [comment]
The travis fail is a timeout error unrelated to the doc changes.

**Comment by phated** [comment]
I'm hoping to review the vinyl-fs docs and other changes Friday night or this weekend.

File: docs/API.md [hunk 43]
```diff
@@ -69,32 +98,83 @@ gulp.src('./client/templates/*.jade')
...
+`cwd` for the output folder, only has an effect if provided output folder is 
+relative.
+
+##### options.mode
+Type: `String`
+
+Default: the mode of the input file (file.stat.mode) or the process mode 
+if the input file has no mode property.
+
+Octal permission string specifying the mode the files should be created with.
```
Comment (phated) [inline docs/API.md]: is this a string?  or is it an actual Octal? `"0777"` vs `0777`

File: docs/API.md [hunk 44]
```diff
@@ -69,32 +98,83 @@ gulp.src('./client/templates/*.jade')
...
+if the input file has no mode property.
+
+Octal permission string specifying the mode the files should be created with. 
+
+##### options.dirMode
+Type: `String`
+
+Default: Default is the process mode.
+
+Octal permission string specifying the mode the directory should be
```
Comment (phated) [inline docs/API.md]: same question as above

File: docs/API.md [hunk 45]
```diff
@@ -69,32 +98,83 @@ gulp.src('./client/templates/*.jade')
...
+
+### gulp.symlink(folder[, options])
+
+Functions exactly like `gulp.dest`, but will create symlinks instead of copying
+a directory.
+
+#### folder
+Type: `String` or `Function`
+
+A folder path or a function that takes in a file and returns a folder path.
```
Comment (phated) [inline docs/API.md]: change `takes` to `receives`

File: docs/API.md [hunk 46]
```diff
@@ -104,133 +184,191 @@ gulp.task('somename', function() {
...
 ```
 
-**Note:** Are your tasks running before the dependencies are complete?  Make sure your dependency tasks are correctly using the async run hints: take in a callback or return a promise or event stream.
+You should either provide the task name or avoid anonymous functions.
 
 #### fn
 
-The function that performs the task's operations. Generally this takes the form of `gulp.src().pipe(someplugin())`.
+The function that performs the task's operations. Generally this takes the 
+form of `return gulp.src().pipe(someplugin());`.
```
Comment (phated) [inline docs/API.md]: this is the function body, not the whole task function.  I think you should make this a full example block of a task function

File: docs/recipes/mocha-test-runner-with-gulp.md [hunk 47]
```diff
@@ -35,6 +35,6 @@ gulp.task('mocha', function() {
 });
 
 gulp.task('watch-mocha', function() {
-    gulp.watch(['lib/**', 'test/**'], ['mocha']);
+    gulp.watch(['lib/**', 'test/**'], 'mocha');
```
Comment (phated) [inline docs/recipes/mocha-test-runner-with-gulp.md]: switch `'mocah'` to `gulp.series('mocha')` just to stay consistent

File: docs/API.md (hunk 43 above)
Comment (dinoboff) [inline docs/API.md]: Tested, and both work.

File: docs/API.md (hunk 44 above)
Comment (dinoboff) [inline docs/API.md]: Tested, and both work.

File: docs/API.md (hunk 46 above)
Comment (dinoboff) [inline docs/API.md]: Like:

    The function that performs the task's operations. Generally it takes this 
    form:
    ```
    gulp.task('someTask', function() {
      return gulp.src(['some/glob/**/*.ext']).pipe(someplugin());
    })
    ```

File: docs/API.md (hunk 43 above)
Comment (dinoboff) [inline docs/API.md]: @phated Should I reword it to:

##### options.mode
Type: `String` or `Number`

Default: the mode of the input file (file.stat.mode) or the process mode if the input file has no mode property.

Octal permission specifying the mode the files should be created with: e.g. `"0777"`, `0777` or `511` (`0777` in base 10).

File: docs/recipes/server-with-livereload-and-css-injection.md [hunk 48]
```diff
@@ -64,7 +64,7 @@ gulp.task('serve', function() {
     }
   });
 
-  gulp.watch('scss/*.scss', ['sass']);
+  gulp.watch('scss/*.scss', 'sass');
```
Comment (phated) [inline docs/recipes/server-with-livereload-and-css-injection.md]: switch `'sass'` to `gulp.series('sass')`

File: docs/recipes/using-external-config-file.md [hunk 49]
```diff
@@ -30,10 +30,11 @@ Beneficial because it's keeping tasks DRY and config.json can be used by another
 ###### `gulpfile.js`
 
 ```js
-// npm install --save-dev gulp gulp-uglify
+// npm install --save-dev gulp gulp-uglify merge-stream
 var gulp = require('gulp');
 var uglify = require('gulp-uglify');
 var config = require('./config.json');
+var merge = require('merge-stream');
```
Comment (phated) [inline docs/recipes/using-external-config-file.md]: this should be `merge2`

File: docs/API.md (hunk 46 above)
Comment (phated) [inline docs/API.md]: yeah

File: docs/API.md (hunk 44 above)
Comment (phated) [inline docs/API.md]: cool, thanks

File: docs/API.md (hunk 43 above)
Comment (phated) [inline docs/API.md]: yeah, i think that would be more clear

**Comment by phated** [comment]
After my comments are addressed, I am :+1:  on this

**Comment by dinoboff** [comment]
@contra can review and merge it please?

File: docs/API.md (hunk 43 above)
Comment (dinoboff) [inline docs/API.md]: done

File: docs/API.md (hunk 45 above)
Comment (dinoboff) [inline docs/API.md]: done

File: docs/API.md (hunk 46 above)
Comment (dinoboff) [inline docs/API.md]: done

File: docs/recipes/mocha-test-runner-with-gulp.md (hunk 47 above)
Comment (dinoboff) [inline docs/recipes/mocha-test-runner-with-gulp.md]: done and it should be consistent.

File: docs/recipes/server-with-livereload-and-css-injection.md (hunk 48 above)
Comment (dinoboff) [inline docs/recipes/server-with-livereload-and-css-injection.md]: done

File: docs/recipes/using-external-config-file.md (hunk 49 above)
Comment (dinoboff) [inline docs/recipes/using-external-config-file.md]: done

File: docs/recipes/using-external-config-file.md [hunk 50]
```diff
@@ -30,10 +30,11 @@ Beneficial because it's keeping tasks DRY and config.json can be used by another
 ###### `gulpfile.js`
 
 ```js
-// npm install --save-dev gulp gulp-uglify
+// npm install --save-dev gulp gulp-uglify merge-stream
```
Comment (phated) [inline docs/recipes/using-external-config-file.md]: missed updating this to merge2

Comment (phated) [inline docs/recipes/using-external-config-file.md]: just checked the amount of downloads on each. I'm going to switch it back to merge-stream

File: docs/recipes/using-external-config-file.md [hunk 51]
```diff
@@ -30,10 +30,11 @@ Beneficial because it's keeping tasks DRY and config.json can be used by another
...
 var gulp = require('gulp');
 var uglify = require('gulp-uglify');
 var config = require('./config.json');
+var merge = require('merge2');
```
Comment (phated) [inline docs/recipes/using-external-config-file.md]: going to switch this back to merge-stream based on download count

**Comment by phated** [comment]
merged

**Comment by Ashwell** [comment]
I have been playing with gulp 4.0, works great for me. In looking at the docs, a function `gulp.symlink` has been documented but apparently not implemented as it is undefined on the gulp instance. Looking at index.html it would seem that below line number 13 this line should be added:

```js
Gulp.prototype.symlink = vfs.symlink;
```

no?

**Comment by phated** [comment]
@Ashwell good catch, just pushed the addition.

**Comment by Ashwell** [comment]
YAY! I helped!
